src/app/util/ember-api.h

```diff
@@ -0,0 +1,348 @@
+/**
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/util/af-types.h>
+#include <app/util/att-storage.h>
+#include <app/util/attribute-metadata.h>
+// #include <app/util/endpoint-config-defines.h>
+// #include <lib/support/CodeUtils.h>
+
+static constexpr uint16_t kEmberInvalidEndpointIndex = 0xFFFF;
+
+/**
+ * @brief locate attribute metadata
+ *
+ * Function returns pointer to the attribute metadata structure,
+ * or NULL if attribute was not found.
+ *
+ * @param endpoint Zigbee endpoint number.
+ * @param clusterId Cluster ID of the sought cluster.
+ * @param attributeId Attribute ID of the sought attribute.
+ *
+ * @return Returns pointer to the attribute metadata location.
+ */
+const EmberAfAttributeMetadata * emberAfLocateAttributeMetadata(chip::EndpointId endpoint, chip::ClusterId clusterId,
+                                                                chip::AttributeId attributeId);
+
+/**
+ * @brief Returns true if endpoint contains the ZCL server with specified id.
+ *
+ * This function returns true if
+ * the endpoint contains server of a given cluster.
+ */
+bool emberAfContainsServer(chip::EndpointId endpoint, chip::ClusterId clusterId);
+
+/**
+ * @brief Returns true if endpoint of given index contains the ZCL server with specified id.
+ *
+ * This function returns true if
+ * the endpoint of given index contains server of a given cluster.
+ * If this function is used with a manufacturer specific clusterId
+ * then this will return the first cluster that it finds in the Cluster table.
+ * and will not return any other clusters that share that id.
+ */
+bool emberAfContainsServerFromIndex(uint16_t index, chip::ClusterId clusterId);
+
+/**
+ * @brief Returns true if endpoint contains the ZCL client with specified id.
+ *
+ * This function returns true if
+ * the endpoint contains client of a given cluster.
+ */
+bool emberAfContainsClient(chip::EndpointId endpoint, chip::ClusterId clusterId);
```

Comment:
this splits the logic of "if xyz is defined in foo.h, it is implemented in foo.cpp". As a result then we have a "ember-api.h" that is available standalone even though the functions that it defines are not linked in and this breaks dependencies.